toolkits/confluence/arcade_confluence/client.py

```diff
@@ -0,0 +1,535 @@
+from enum import Enum
+from typing import Any
+from urllib.parse import parse_qs, urlparse
+
+import httpx
+from arcade.sdk.errors import ToolExecutionError
+
+from arcade_confluence.enums import BodyFormat, PageUpdateMode
+from arcade_confluence.utils import (
+    build_child_url,
+    build_hierarchy,
+    remove_none_values,
+    split_by_space,
+)
+
+
+class ConfluenceAPIVersion(str, Enum):
+    V1 = "wiki/rest/api"
+    V2 = "wiki/api/v2"
+
+
+class ConfluenceClient:
+    ACCESSIBLE_RESOURCES_URL = "https://api.atlassian.com/oauth/token/accessible-resources"
+    BASE_URL = "https://api.atlassian.com/ex/confluence"
+
+    def __init__(self, token: str, api_version: ConfluenceAPIVersion):
+        self.token = token
+        self.cloud_id = self._get_cloud_id()
+        self.api_version = api_version.value
+
+    def _get_cloud_id(self) -> str:
+        """
+        Fetch the cloudId for <workspace_name>.atlassian.net
+        using the OAuth2 3LO accessible-resources endpoint.
+
+        For details on why this is necessary, see: https://developer.atlassian.com/cloud/oauth/getting-started/making-calls-to-api
+        """
+        headers = {"Authorization": f"Bearer {self.token}"}
+        resp = httpx.get(self.ACCESSIBLE_RESOURCES_URL, headers=headers)
+        resp.raise_for_status()
+        resp_json = resp.json()
+
+        if len(resp_json) == 0:
+            raise ToolExecutionError(message="No workspaces found for the authenticated user.")
+
+        return resp_json[0].get("id")  # type: ignore[no-any-return]
+
+    async def request(self, method: str, path: str, **kwargs: Any) -> Any:
+        headers = {
+            "Accept": "application/json",
+            "Authorization": f"Bearer {self.token}",
+        }
+        async with httpx.AsyncClient() as client:
+            response = await client.request(
+                method,
+                f"{self.BASE_URL}/{self.cloud_id}/{self.api_version}/{path.lstrip('/')}",
+                headers=headers,
+                **kwargs,
+            )
+            response.raise_for_status()
+            return response.json()
+
+    async def get(self, path: str, **kwargs: Any) -> Any:
+        return await self.request("GET", path, **kwargs)
+
+    async def post(self, path: str, **kwargs: Any) -> Any:
+        return await self.request("POST", path, **kwargs)
+
+    async def put(self, path: str, **kwargs: Any) -> Any:
+        return await self.request("PUT", path, **kwargs)
+
+
+class ConfluenceClientV1(ConfluenceClient):
+    def __init__(self, token: str):
+        super().__init__(token, api_version=ConfluenceAPIVersion.V1)
+
+    def construct_cql(
```

Comment:
Should constructing CQL strings be a part of the HTTP client? I feel like this belongs to some utilities module.

Reply:
Going to keep ALL logic related to the V1 API encapsulated in the same class so that refactoring is easy in the future when the V12 search API is deprecated.